bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -91,87 +91,162 @@
-                    public Object addingService(ServiceReference reference) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
-                        return provider;
-                    }
+    /**
+     * Activates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     * @throws Exception
+     */
+    protected void activate(ComponentContext componentContext) throws Exception {
+        this.bundleContext = componentContext.getBundleContext();
+        this.ruleEngine = new RuleEngine(bundleContext, moduleTypeRegistry);
+        this.ruleEngine.setStatusInfoCallback(this);
+        super.activate(bundleContext);
+    }
 
-                    @Override
-                    public void modifiedService(ServiceReference reference, Object service) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
+    /**
+     * Deactivates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     */
+    protected void deactivate(ComponentContext componentContext) {
+        super.deactivate();
+        // removeManagedProvider(managedProvider);
+        ruleEngine.dispose();
+        ruleEngine = null;
+    }
 
-                    }
+    /**
+     * Bind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void setModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = moduleTypeRegistry;
+    }
 
-                    @Override
-                    public void removedService(ServiceReference reference, Object service) {
-                    }
+    /**
+     * Unbind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void unsetModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = null;
+    }
 
-                });
-        templateProviderTracker.open();
+    /**
+     * Bind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void setTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = templateRegistry;
+    }
 
+    /**
+     * Unbind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void unsetTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = null;
+    }
+
+    protected void setStorageService(StorageService storageService) {
+        // this.storageService = storageService;
```

Comment:
The RuleRegistry is using RuleEngine in its implementation. How the RuleRegistry will get the RuleEngine component without public interface? For me the RuleEngine must not be component at all when having 1:1 relation between RuleRegistry and RuleEngine.
RuleRegistry must have dynamic policy to ManagedProvider. (RuleRegistry must not be deactivated when the storage is disappeared)
Our proposal is that:
1) RuleRegistry to be a component
2) RuleRegistry has dynamic relation to ManageProvider 
3) RuleRegistry creates a RuleEngine (internally, it will not be a component - it is part of the RuleRegistry and they will have 1:1 dependency)
There is one problem that the RuleRegistry can be available without ManagedProvider and all RuleRegistry.add method will throws exception. The applications have to listen for ManagedRuleProvider instead of RuleRegistry to add (update, remove) rules.